Fix issues #2728 and #2733

maclomhair opened the pull request:
Issue #2728 is because the tokenizer was set to remove diacritics on words, but not in the search query. To be closer to the expected, traditional behavior of the search, I disabled the removal of diacritics now and introduced a change to detect if the full-text table schema is outdated, in which case the tables are recreated. This should fix the issue for everyone without having to delete the entire index.db.

Issue #2733 stems from a misunderstanding about the sqlite full-text search query language when I implemented the plugin. The query language doesn't actually support proper globbing. Instead, I now simulate the globbing by first globbing through all terms known to the index and constructing a query from that, then searching with that newly constructed query instead. I haven't done performance measurements on this approach but I expect that it should be unnoticable fast in general. A potential speedup might otherwise be to introduce an index on all tokens known to sqlite FTS.

I'm looking forward to the review and I'm happy to incorporate the review findings!

maclomhair committed:
Fix issue #2728 properly by instructing tokenizer to not remove diacritics

maclomhair committed:
Fix issue #2733 to add proper globbing support to indexed FTS

This works by performing the globbing on all tokens, then searching for
an ORed combination of all tokens in the full-text index. I hope this is
sufficiently fast to not be noticeable.

Further optimization might be possible by creating an index over the list
of tokens.

jaap-karssenberg commented:
I can't really contribute review here, as I'm not familiar with this SQLite module. But happy to merge as is, when you think the patch is good to go.

--- Jaap

maclomhair commented:
Ok! I'll see if I can get the bug reporters to test that the changes fix the issues for them to make sure... maybe I can get some more feedback. If that doesn't work out - I've looked over the changes again now and I'm confident that they do fix the issues.

roland-ruedenauer commented:
Hi @maclomhair,

I did some testing on Windows using your latest changes and found global search to work really well.
Search terms containing (german) umlauts is working for me and wildcard search using "*" too.

Some minor issues (not directly related to #2728 or #2733) I found

* a search term ending in a quotation mark can trigger an error if the prefix is short (e.g. "ref?")
* searching for phrases using double quotes as mentioned in the manual doesn't seem to be working - e.g. searching for "C++20"
* accidently searching for "*" took really long and I wasn't able to cancel the search via the cancel button (I finally had to kill the zim process)
* using && or || operators as mentioned in the manual does not work (also without indexed search pluging, I guess)

sojusnik commented:
@maclomhair Can you please check if your updates also fix searching for words containing hyphens, like in lorem-ipsum, since they aren't found in your initial version?

Happens on Ubuntu 24.04 with Zim 0.76.0.

maclomhair commented:
Thanks for your testing @roland-ruedenauer and your hints @sojusnik!

> * a search term ending in a quotation mark can trigger an error if the prefix is short (e.g. "ref?")

I'll fix this in the next revision of this pull request! Thanks for pointing it out!

> * searching for phrases using double quotes as mentioned in the manual doesn't seem to be working - e.g. searching for "C++20"

My testing showed that using double quotes in search works normally but the problem here is that the full-text search sees "+" as a separator token, so it indexes only "C" and "20" instead. It is possible to specify custom mapping for which characters are tokens and which are separators, and I think making the mapping customizable is the way to go here.

> * accidently searching for "*" took really long and I wasn't able to cancel the search via the cancel button (I finally had to kill the zim process)

I'll fix this by adding a fast return if mistakenly searching for "*", so that the search immediately terminates without results instead.

> * using && or || operators as mentioned in the manual does not work (also without indexed search pluging, I guess)

This seems to be a search bug, not related to FTS. Using "AND" or "OR" instead seems to work, though!

> Can you please check if your updates also fix searching for words containing hyphens, like in lorem-ipsum, since they aren't found in your initial version?

This again is because "lorem-ipsum" is by default split into "lorem" and "ipsum"... I plan to add an option where people can add custom letters which should be considered tokens to be indexed.

jaap-karssenberg commented:
If nothing else comes up, I'll merge these fixes on Thursday evening and push out a new release. Next "point" release will be roughly a month later, so further improvements can land there.

maclomhair committed:
Add configuration options to full-text search, incl. help text in manual

maclomhair committed:
Merge branch 'develop' into indexed_fts_fixup

maclomhair commented:
I have now reworked the fixes a little bit to give users more options to fine-tune their full-text indexing. 

Specificly, I have added options to enable/disable the removal of diacritics, which by default is turned off to better resemble how traditional search works.

A second option that can be globally configured is a string where additional characters which are considered tokens (i.e. part of a word) can be specified. Adding "+" there, for example, will lead to strings such as "C++20" being indexed as one string.

I have also added a help text for both options which can be found in the manual.

The plugin remembers which options were set when the index was last created, and updating the options will automatically lead to the recreation of the full-text search index should it be necessary.

I think this approach is overall better suited for the different user needs regarding full-text search. I've tested the code a couple of times and I think it's good to go.

(Ah and of course, I've also added a short return path if accidentally searching for "*", and "?" should be properly escaped now)
